pkg/kubelet/eviction/threshold_notifier_linux.go

```diff
@@ -67,51 +100,105 @@ func NewMemCGThresholdNotifier(path, attribute, threshold, description string, h
 			unix.Close(eventfd)
 		}
 	}()
-	glog.V(3).Infof("eviction: setting notification threshold to %s", threshold)
-	config := fmt.Sprintf("%d %d %s", eventfd, watchfd, threshold)
+	config := fmt.Sprintf("%d %d %d", eventfd, watchfd, threshold)
 	_, err = unix.Write(controlfd, []byte(config))
 	if err != nil {
 		return nil, err
 	}
-	return &memcgThresholdNotifier{
-		watchfd:     watchfd,
-		controlfd:   controlfd,
-		eventfd:     eventfd,
-		handler:     handler,
-		description: description,
-	}, nil
+	notifier := &cgroupNotifier{
+		watchfd:   watchfd,
+		controlfd: controlfd,
+		eventfd:   eventfd,
+		stop:      make(chan struct{}),
+	}
+	go notifier.getThresholdEvents(eventCh)
+	return notifier, nil
 }
 
-func getThresholdEvents(eventfd int, eventCh chan<- struct{}, stop ThresholdStopCh) {
+// getThresholdEvents waits for events on the eventfd and passes a struct into the eventCh when an event occurrs.
+func (n *cgroupNotifier) getThresholdEvents(eventCh chan<- struct{}) {
+	epfd, err := unix.EpollCreate1(0)
+	if epfd == -1 {
+		glog.Warningf("eviction manager: error creating epoll: %v", err)
+		return
+	}
+	event := unix.EpollEvent{
+		Fd:     int32(n.eventfd),
+		Events: unix.EPOLLIN,
+	}
+	err = unix.EpollCtl(epfd, unix.EPOLL_CTL_ADD, n.eventfd, &event)
+	if err != nil {
+		glog.Warningf("eviction manager: error adding epoll eventfd: %v", err)
+		return
+	}
 	for {
-		buf := make([]byte, 8)
-		_, err := unix.Read(eventfd, buf)
+		event, err := wait(epfd, n.eventfd)
 		if err != nil {
+			glog.Warningf("eviction manager: error while waiting for memcg events: %v", err)
 			return
+		} else if event {
+			// Consume the event from the eventfd
+			buf := make([]byte, 8)
```

Comment:
Is there a constant somewhere in the standard lib that describes how big this buffer should be?

Reply:
not sure.  This is part of the current implementation: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/threshold_notifier_linux.go#L87

Reply:
This is part of the Linux eventfd API. See [the man page](http://man7.org/linux/man-pages/man2/eventfd.2.html), which says:

```
       read(2)
              Each successful read(2) returns an 8-byte integer.  A read(2)
              fails with the error EINVAL if the size of the supplied buffer
              is less than 8 bytes.
```

I'm not sure a constant exists for this size... It doesn't seem to exist in the C API either...

Reply:
doen